### DISCUSSION TITLE
Using Telerik Blazor UI

### DISCUSSION THREAD

**Discussion by skuflow** [oqtane/oqtane.framework#3037] [Q&A]
Is it feasible to use Telerik Blazor UI in custom modules?  I generated a module using the oqtane admin but there is no index.html page in into which we can hook the js references.  Please let me know asap.  This may be a deal killer for us.

**Comment by sbwalker** [comment]
If you search in the Discussions you will find references to using third party component libraries in Oqtane. Note that in a SPA architecture like Blazor there is only a single "page" on the server - _Host.cshtml. Obviously you do not want to modify this page as it is part of the framework so you need to leverage the extensibility points available. @nicpitsch is using Telerik in Oqtane... perhaps he can comment.  Note that in Oqtane 4.0 the IHostResources interface was deprecated so you need to use the new  methods to register site-level JavaScript (if required) - documented here: https://www.oqtane.org/blog/!/61/css-js-resources

**Reply by skuflow** [reply]
it looks like oqtane supports both webassembly or webserver for Blazor.  If we choose web assembly, modules should be still a spa and not setup with an @page configuration, correct?  The idea is to write modules and then drop them on an oqtane page?

**Reply by sbwalker** [reply]
@skuflow the @page attribute has nothing to do with the Blazor hosting model - it is simply syntactic sugar which Microsoft included to work with the default Blazor router. Oqtane does not use the default Blazor router - it uses a custom router... so the @page attribute is ignored. You are correct that Oqtane supports both Server and WebAssembly hosting models.

**Reply by skuflow** [reply]
where do we add builder.Services.AddTelerikBlazor();?   there's no program.cs file in a module

**Reply by skuflow** [reply]
there's also no /shared/mainlayout.razor page?

**Reply by sbwalker** [reply]
Correct - Oqtane is a framework not an application template (https://media.licdn.com/dms/image/D4E22AQES7z1qHtDnaw/feedshare-shrink_800/0/1688558359918?e=1692230400&v=beta&t=cMbShA-9NlQ5MkOJDraEMniI-LvexjtsW-biuDxi-Gw)... in a framework you use extensibility points to extend the functionality rather than modifying existing code files. I just searched in the Discussions area of this repo for "Telerik" and I found this item #690 which is titled "Register Telerik UI for Blazor"

**Comment by nicpitsch** [comment]
I have created the github repo https://github.com/nicpitsch/Oqtane.TelerikModule1 documenting the necessary steps to integrate *Telerik Blazor UI Components* into Oqtane 4.x.

**Reply by sbwalker** [reply]
Thank you @nicpitsch - this is very helpful!

**Reply by skuflow** [reply]
what is the best way to get to the telerik nuget packages?  Our team has them all being put in c:\Users\<UserName>\.nuget\ globally so this means that they will have to update their username in this path if they are working on the module locally, correct? I'm trying this:

XCOPY "C:\Users\<UserNameHere>\.nuget\packages\telerik.ui.for.blazor\4.4.0\Telerik*.dll" "..\..\oqtane.framework\Oqtane.Server\bin\Debug\net7.0\" /Y

XCOPY "..\Client\bin\Debug\net7.0\win10-x64\publish\wwwroot\_content\Telerik.UI.for.Blazor\*" "..\..\oqtane.framework\Oqtane.Server\wwwroot\Modules\My.Module.TelerikModule1\Telerik.UI.for.Blazor\" /Y /S /I


I can't even find the _content directory in the second statement anywhere.

**Reply by nicpitsch** [reply]
> what is the best way to get to the telerik nuget packages? 

@skuflow, as explained in chapter [*Install Telerik.Blazor NuGet package*](https://github.com/nicpitsch/Oqtane.TelerikModule1#install-telerikblazor-nuget-package) in my github repo I did it using `dotnet publish --configuration Debug -p:PublishSingleFile=false -p:PublishTrimmed=false --runtime win10-x64 --self-contained false`.  This is when the *.\Client\bin\Debug\net7.0\win10-x64\publish* folder structure is created. The *publish* folder itself contains the **dll's** and the *.\wwwroot\_content\Telerik.UI.for.Blazor* subfolder contains the **css and js**:

![publish folder content](https://github.com/oqtane/oqtane.framework/assets/1652835/6408b6c8-ea18-4aba-b83e-04a6562d020c)

On the other hand using `<CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>` as explained by @sbwalker, refering [Oqtane Module Development Chapter7 - YouTube](https://youtu.be/eoZ7NbtJ-ak?t=300), put's the Telerik **dll's** into place. With this option you won't get the css and js.

**Comment by leigh-pointer** [comment]
@skuflow 

> I can't even find the _content directory in the second statement anywhere.

I think you are missing a backslash 

`XCOPY "..\Client\bin\Debug\net7.0\win10-x64\publish\wwwroot\_content\Telerik.UI.for.Blazor*" "....\oqtane.framework\Oqtane.Server\wwwroot\Modules\My.Module.TelerikModule1\Telerik.UI.for.Blazor" /Y /S /I`

**Reply by skuflow** [reply]
well it's more than that i think... there is no _content directory, even a hidden one and the telerik nuget package on the first xcopy line can't be found.

**Reply by skuflow** [reply]
ie this path literally doesn't exist (C:\source\repos\PDI.Module.Report\Client\bin\Debug\net7.0\wwwroot\_content)

**Reply by sbwalker** [reply]
@skuflow Have you tried adding ```<CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>``` to your project file? I believe this setting used to be included in the external module template. It extracts assemblies from nuget packages.

**Reply by skuflow** [reply]
i don't see it in the Client project file.   There is no "external module" template in version 4 that I see.... there's just the default module template

**Reply by sbwalker** [reply]
sorry... you are correct, it is called "Default Module Template" now. 

CopyLocalLockFileAssemblies

The CopyLocalLockFileAssemblies property is useful for plugin projects that have dependencies on other libraries. If you set this property to true, any NuGet package dependencies are copied to the output directory. That means you can use the output of dotnet build to run your plugin on any machine.

**Reply by skuflow** [reply]
If i add that the the csproj file do I need to do the xcopy stuff too?  How does it know where to put stuff?

**Reply by leigh-pointer** [reply]
@skuflow 
In the packages folder is a debug.cmd file open that and it should give you some direction

**Reply by skuflow** [reply]
right but if you scroll up the debug.cmd file can't even find the paths i'm referencing.   

This command:

XCOPY "C:\Users\<usernamehere>\.nuget\packages\telerik.ui.for.blazor\4.4.0\\_content\\*" "..\..\oqtane.framework\Oqtane.Server\wwwroot\Modules\PDI.Module.Report\Telerik.UI.for.Blazor\"

results in this output:

5>File not found - *
5>C:\source\repos\PDI.Module.Report\Package\PDI.Module.Report.Package.csproj(22,5): error MSB3073: The command "debug.cmd" exited with code 4.

there is no _content directory in the path for the telerik nuget package.

**Reply by sbwalker** [reply]
If you add the CopyLocalLockFileAssemblies node to your project file and build the project, the assemblies will be extracted from the Nuget package and copied to the local /bin with your own module assemblies. Then you can update your debug.cmd paths to copy the assemblies from your local /bin to the Oqtane framework /bin.

**Reply by sbwalker** [reply]
I believe this is explained in the module development video tutorial - https://www.youtube.com/watch?v=JPfUZPlRRCE&list=PLYhXmd7yV0elLNLfQwZBUlM7ZSMYPTZ_f
